Show volume on the gas analyzer

slarticodefast opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
The gas analyzer now shows the volume of pipes, tanks, canisters and the environment. The volume always corresponds to the moles listed. ~~So for pipe networks the total volume is shown, for the environment the volume of the tile the player is in.~~ For pipe networks the volume and the moles of the analyzed pipe element will be shown.
## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
The volume of tiles, pipes, tanks and canisters is not mentioned ingame. While it can be calculated in theory by using pV=nRT for each entity and memorizing it, having all variables in this formula shown should make atmos more accessible to new players and makes it possible to verify the formula.

Also, did you know that wrenching a canister to the connector port extends its volume by one liter? Now you do :)
## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
Adds an additional field to the GasMixEntry struct for the volume.
Adds a new line to the GasAnalyzerWindow.
Adds two new translation strings for the volume.

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

https://github.com/space-wizards/space-station-14/assets/161409025/fa1582ea-d10a-4c24-a8e0-82c34892da38



- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->
The GasMixEntry struct has an additional field for the volume. This is only used for the gas analyzer and has been adjusted for all constructor calls.
Two new localization strings must be added and translated.
GasAnalyzerScanEvent now uses a list instead of a dictionary.
**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

:cl:
- add: The gas analyzer now tells you the volume of scanned objects. Tanks inside canisters now show up as well.
- tweak: The gas analyzer now shows the amount of moles inside the scanned pipe element instead of the whole pipe network.

deltanedas commented:
how is the gas analyzer able to know how many moles are in the entire distro network?

Partmedia commented:
> how is the gas analyzer able to know how many moles are in the entire distro network?

Not sure, maybe it should be a pressure-only sensor. You're not contributing meaningfully to the discussion with a one-sentence response like this.

Varen commented:
> Why does this make it more "accessible?" Do you have any evidence for this? What would you do with the volume if it were reported to you?

I have one point on accessibility: When I look at a bottle in real life, I can tell if it's about half a liter, a liter, two liters and so on. We don't have the ability to tell at a glance how big something is comparatively in this game, and in many cases it's not quite intuitive: 200L for a pipe piece, but on screen it looks about as big as a 5L tank. Also a tile can contain 2500L? That's neat, I now learnt that our tiles are officially 1.35721m long. That is a cursed size if I've ever seen one.

All this to say: I'd have the volume on examine, but it'd probably get cumbersome, so having it in the gas analyzer is probably good enough. 

> Can you propose a quasi-scientific explanation for how this can be measured?
I feel like this goes a bit into lore over game mechanics, but uh, call it standardized NT given RFID chips if nothing comes to mind as to why. Or bluespace echo magic thing technology?

Partmedia commented:
> I have one point on accessibility: When I look at a bottle in real life, I can tell if it's about half a liter, a liter, two liters and so on. 

Then this should only work for bottles, and not for entire pipe networks.

Can you just justify why your real life example should also apply to the whole pipenet that is hidden under floors and possibly snaking around the whole station?

Varen commented:
> > I have one point on accessibility: When I look at a bottle in real life, I can tell if it's about half a liter, a liter, two liters and so on.
> 
> Then this should only work for bottles, and not for entire pipe networks.
> 
> Can you just justify why your real life example should also apply to the whole pipenet that is hidden under floors and possibly snaking around the whole station?

Chromiumboy is working on an atmos monitoring console that will give us a view of the whole network on the map, with views of the single atmospherics alarms and the likes. If we can justify the tech on station to do that, then getting the connected pipecount * 200 on a handheld device is not really a difficult task through the same systems that see that information.

Partmedia commented:
> an atmos monitoring console that will give us a view of the whole network on the map... then getting the connected pipecount * 200 on a handheld device is not really a difficult task through the same systems that see that information.

This is not really relevant to this discussion because:

1. The atmos monitoring console is anchored to the ground, not something you carry in your hand
2. Then return the `pipecount * 200` on a handheld device, and not the real volume

Furthermore, I would like to see more justification how and why this would benefit "accessibility", and why the solution I proposed would not work.

Ilya246 commented:
as long as the gas analyzer shows gas mols this is literally just QoL as you can calculate volume as `V=nRT/P`

Varen commented:
> > an atmos monitoring console that will give us a view of the whole network on the map... then getting the connected pipecount * 200 on a handheld device is not really a difficult task through the same systems that see that information.
> 
> This is not really relevant to this discussion because:
> 
>     1. The atmos monitoring console is anchored to the ground, not something you carry in your hand
> 
>     2. Then return the `pipecount * 200` on a handheld device, and not the real volume

Honestly, I can think of a million ways this can work, given the prevalence of wireless information transfer and all kinds of API that must exist in world. A full specific on why is getting lost in the weeds on a PR like this, however, as it seems plausible enough.

> Furthermore, I would like to see more justification how and why this would benefit "accessibility", and why the solution I proposed would not work.
Requiring the player to do math to know a variable that tells them how big a container is should have a really good design decision, and right now I don't see one. It's not a detail that creates interesting design choices or decisions, just the hassle to whip up a calculator and make a cheatsheet of volume for each object type, or memorization.

Making the gas analyzer calculate it clientside works just fine, in this context. I don't see the need to do it on an "upgraded" gas analyzer however, for the reasons stated above.

slarticodefast commented:
> Can you propose a quasi-scientific explanation for how this can be measured?
Even with our current technogy we can convert [convert videos to 3d models](https://www.adobe.com/products/substance3d/discover/create-3d-models-from-video.html). Add something like an x-ray scanner and you can look at the inside of a pipe. Or simpler, let's say all pipes and tank come with an RFID chip which tells you its volume.

> how is the gas analyzer able to know how many moles are in the entire distro network?
I think at some point we can ignore some realism for better QoL. How do the chem goggles work? It is science fiction after all.

>Why does this make it more "accessible?"
Imagine you are new to atmos and not good at physics. You can look up the formula $pV=nRT$ in the guidebook. IF you want to find the volume of a pipe/tank/canister, you have to:
1. Realize you got the variables V,nT and the constant R and can solve for V.
2. Fill the pipe with some gas and measure the pressure and moles.
3. Write the result down somewhere or remember it.
4. Repeat this for all different pipe elements, because unary ones got different volume than binary or trinary ones.
5. Hope the result is correct, because you can't really verify it.

This becomes even more complicated for mixtures of ideal gases, because you have to know about partial pressures.

If you are not an atmos player and just want to use a gas tank for internals, how much is the regular tank better than the emergency one? What about the double emergency tank?

> Making the gas analyzer calculate it clientside works just fine, in this context.
I think sending one extra already calculated floating point value from the server to the client might be more efficient than adding up the partial pressures clientside. There are usually not that many gas analyzers running at the same time.

Partmedia commented:
> Honestly, I can think of a million ways this can work, given the prevalence of wireless information transfer and all kinds of API that must exist in world.

This argument would seem to justify an "are you an antagonist" HUD.

> > Requiring the player to do math to know a variable that tells them how big a container is should have a really good design decision, and right now I don't see one.

This implies that not having this feature "[requires] the player to do math" which is not true. Can you show how the current implementation requires the player to do math? Don't set up a [straw man, then knock down the straw man](https://en.wikipedia.org/wiki/Straw_man).

Your premise is:

> If you want to find the volume of a pipe/tank/canister,

But your premise may not be true, because when does an actual player need to find the volume?

> If you are not an atmos player and just want to use a gas tank for internals, how much is the regular tank better than the emergency one? What about the double emergency tank?

You fill it up, then see which one has more moles. None of this requires looking at the volume. In fact, this is more engaging for gameplay than running a scanner on a container because then you would have to go fill it up and find out.

Would you prefer a game where players *do things* to find out about the game, or where a player can just armchair wiki all the numbers out?

slarticodefast commented:
I got one more example why this is useful: The gas analyzer does not show where the moles are located. So if you scan a pipe, are only the moles inside this pipe element shown or those of the whole network? Showing the volume clarifies this.
Same thing for the moles in the environment. Is the moles in the tile, per liter or in the whole room?

One could even argue that the volume, temperature and pressure should be shown and not the moles. The volume is a fundamental property of the container and the pressure and temperature are measured by sensors. The moles are the calculated quantity.

> Would you prefer a game where players do things to find out about the game, or where a player can just armchair wiki all the numbers out?

Personally I would argue that the fun lies in actually designing pipe setups and less in having to figure out how the parts work. You are of course right that this is a design decision about QoL, which is not mine to make.

Golinth commented:
> Would you prefer a game where players _do things_ to find out about the game, or where a player can just armchair wiki all the numbers out?

Players already do armchair wiki all the numbers out in atmos, this just makes this information easily available in game, with exact amounts rather than guesswork or relying on others. Not to mention, all of these volume calculations already only take a one time calculation to figure out. "Forcing" curious atmos players to do the calculation themselves once for a spreadsheet to reference is not engaging gameplay, its tedium. This moves away from that approach, and I 100% support it.

Partmedia commented:
As discussed on Discord, this could be bad for accessibility because it leads to confusing measurements of volume.

For pipenets, it shows the volume of the entire pipenet, even if the rest of it is not visible:

![image](https://github.com/space-wizards/space-station-14/assets/3229565/313abdaa-3700-4f75-93aa-74e64355d33e)

For the environment, it shows the volume of only the current tile:

![image](https://github.com/space-wizards/space-station-14/assets/3229565/909b9a1b-72d6-4451-8d0c-3faed37548e8)

This could lead to the mistaken understanding that all rooms hold the same amount of air.

However, since this was a bug carried over from the existing gas analyzer, I'm willing to consider this PR in its own right.

> only scenario where i see this being useful is calculating max burn energy for a chamber

I want to avoid mechanics that are only useful for setting up burn chambers. However, a few others on Discord have probably convinced me that this is something that is needed and useful. Furthermore, it really is data that is already available.

Since all of the state information is already available, I think it would be best for the client to calculate this number locally rather than the server sending redundant information over the network.

deltanedas commented:
the confusion between pipenet/atmosphere using the entire network vs a single cell is an argument for just showing pressure and %/ppm. and even then, everyone knows that gas analyzer shows the current tile you are on and not the entire room's total mole count

slarticodefast commented:
>For pipenets, it shows the volume of the entire pipenet, even if the rest of it is not visible.

Yes, but this is the same for moles. Having the corresponding volume clarifies where the moles are located.
It would propably be better to only show the moles AND volume of a single pipe element for consistency.

UbaserB commented:
can you PLEASE make the L not lowercase? Hurts my soul..

slarticodefast commented:
>can you PLEASE make the L not lowercase? Hurts my soul..

Ok, I just checked Wikipedia because i have never seen anyone use a capital L for liter and I'm a physicist.  Seems like it is an US thing. Usually capital letters are used for units named after a person. Other opinions on what to use? Would be more readable for sure.

>There are two international standard symbols for the litre: L and l. In the United States the former is preferred because of the risk that (in some fonts) the letter l and the digit 1 may be confused.

Partmedia commented:
> Ok, I just checked Wikipedia because i have never seen anyone use a capital L for liter and I'm a physicist. Seems like it is an US thing. Usually capital letters are used for units named after a person. Anyone else who got an opinion about this? Would be more readable for sure. Do we have the unit anywhere else in the game?

You handled this correctly by adding a FTL localization for this. For en-US you can use capital L (e.g. 4 L). Other locales can choose to adapt this as needed. Thanks.

UbaserB commented:
> > can you PLEASE make the L not lowercase? Hurts my soul..
> 
> Ok, I just checked Wikipedia because i have never seen anyone use a capital L for liter and I'm a physicist. Seems like it is an US thing. Usually capital letters are used for units named after a person. Anyone else who got an opinion about this? Would be more readable for sure. Do we have the unit anywhere else in the game?
> 
> > There are two international standard symbols for the litre: L and l. In the United States the former is preferred because of the risk that (in some fonts) the letter l and the digit 1 may be confused.

im literally learning chemistry and when we're doing calculations based on SLC or conversion of moles we use L

slarticodefast commented:
> You handled this correctly by adding a FTL localization for this. For en-US you can use capital L (e.g. 4 L). Other locales can choose to adapt this as needed. Thanks.

Yeah, for en-US this should be a capital L. I will change this tomorrow.

> im literally learning chemistry and when we're doing calculations based on SLC or conversion of moles <-> volume we use L

I guess this differs a lot by country and branch of science. Thanks for the answer.

deltanedas commented:
L is better since its easier to read

Ilya246 commented:
i think being able to know volume and mols of pipes is good despite the magicality of it
if you were unable to see both moles and volume, that would only promote more codecrawling and moderately annoying math: you would still be able to figure out the volume of pipenets, but you would need to either memorise the amount of pipe nodes (works for prebuilt station nets) or count the pipes you built then multiply by the volume value you can only know from codecrawling or a rather complicated experiment (change mols or volume by a known value and see how the pressure changes, but this still requires codecrawling for your first ever measurement), thus making atmos even more complicated unless you "just know" things and numbers or know how to codecrawl

slarticodefast commented:
The unit symbol has been changed to L.

>Since all of the state information is already available, I think it would be best for the client to calculate this number locally rather than the server sending redundant information over the network.

I had a look at the code again to implement this. However, since there can be a volume without any gas and 0 pressure, the volume cannot be calculated in all cases. Instead one could transfer the volume, temperature and moles to the client and calculate the pressure locally. This creates another problem though: the gas analyzer does not show gases with less than 0.01 moles.

![code 1](https://github.com/space-wizards/space-station-14/assets/161409025/5be00699-7090-4d33-b889-53cb0c99db64)
![code 2](https://github.com/space-wizards/space-station-14/assets/161409025/a02a1f5d-8cb2-4840-b84f-5c5a38c1bd41)
![gas analyzer upstream](https://github.com/space-wizards/space-station-14/assets/161409025/525e0580-f913-41f5-91ba-d7d223d620da)
The ingame screenshot is taken from upstream.
Possible solutions:
1. Leave this PR as it is. This means the pressure and volume are both send to the client and diplayed correctly. The moles under the threshold are not shown, just as before.
2. Calculate the pressure from the volume, temperature and moles locally. However this will lead to 2 different values for the pressure in some cases. In the above example the gas analyzer will show 0kPa while the tank itself shows 123kPa.
3. Like option 2, but remove the threshold for the gas analyzer and have it show trace amounts of gas. Might cause UI problems. The pressure calculation would be accurate.

Partmedia commented:
Thank you for investigating the changes that would be necessary.

I was under the impression that "volume... can be calculated in theory by using pV=nRT" (@slarticodefast) and that this is "this is literally just QoL" (@Ilya246) but it looks like this isn't the case because to implement this the gas analyzer would need volume information from the server.

For option (1), I feel that gas analyzers are already strong enough and don't need any more buffing. I think it would be okay somewhere else, e.g. an upgrade as an "advanced gas analyzer" or a scan mode that uses a longer doafter to complete. Players would have to trade off having technology or spending more time scanning for more information, instead of always getting all the information instantly.

As discussed in the context of health and plant analyzers, it's not good to have a scanner that always gives you all the information instantly. Please let me know if you have trouble finding the discussion on this.

For option (2), did you investigate why it would "lead to 2 different values for the pressure in some cases" and if there's a way to fix it?

I don't think option (3) is viable because trace amounts of gas below the number of digits displayed typically always remain despite having no effect in reactions. You'll see from the git log that this was a deliberate decision made.

> If you are not an atmos player and just want to use a gas tank for internals, how much is the regular tank better than the emergency one? What about the double emergency tank?

I feel that this is better accomplished by putting the container size in the examine text. Please feel free to make a separate PR for this.

Ilya246 commented:
> For option (1), I feel that gas analyzers are already strong enough and don't need any more buffing. I think it would be okay somewhere else, e.g. an upgrade as an "advanced gas analyzer" or a scan mode that uses a longer doafter to complete. Players would have to trade off having technology or spending more time scanning for more information, instead of always getting all the information instantly.

except, as already dicussed, this is information you already effectively know, this PR just reduces the need for doing math and/or codecrawling to know volumes of things
a plant analyzer would be different in that the scanner lets you know information you would otherwise have to do additional ingame actions to find out, here all you have to do is just math and/or codecrawling for info

Partmedia commented:
> except, as already dicussed, this is information you already effectively know, this PR just reduces the need for doing math and/or codecrawling to know volumes of things a plant analyzer would be different in that such a scanner would let you know information you would otherwise have to do additional ingame actions to find out, here all you have to do is just math and/or codecrawling for info, which should be minimised

Please read, namely:

> However, since there can be a volume without any gas and 0 pressure, the volume cannot be calculated in all cases.

slarticodefast commented:
> "volume... can be calculated in theory by using pV=nRT"

You can, but there has to be gas in the container. Otherwise the pressure is 0 and you cannot solve for V. The volume is in my opinion information that should easily be available to the player. It is more intuitive than the number of moles (and not temperature dependent) and in real life each gas container will have its volume just printed on it. The alternative would be showing the volume by inspecting it, but I think you do not need that information all the time as it does not change like the pressure.

Having the volume of a pipe network shown additionally teaches new players if pipes are connected properly or not.

>For option (2), did you investigate why it would "lead to 2 different values for the pressure in some cases" and if there's a way to fix it?

This is shown in the given code snippets and pictures. In this example the container contains 0.0098mol of gas. Gases are only sent to the client if more than 0.01mol are present. Therefore the gas analyzer shows 0mol of gas. No gas means no pressure if calculated locally. If the pressure is sent to the client seperately from the server, all gases are included, even at small amounts.

I also don't think option 3 is viable. I think option 1 (the current state of this PR) is the most consistent one, even if the server has to sent a floating point value more.

Ilya246 commented:
> > except, as already dicussed, this is information you already effectively know, this PR just reduces the need for doing math and/or codecrawling to know volumes of things a plant analyzer would be different in that such a scanner would let you know information you would otherwise have to do additional ingame actions to find out, here all you have to do is just math and/or codecrawling for info, which should be minimised
> 
> Please read, namely:
> 
> > However, since there can be a volume without any gas and 0 pressure, the volume cannot be calculated in all cases.

it could also instead calculate pressure from knowing the volume, temp and moles

slarticodefast commented:
> it could also instead calculate pressure from knowing the volume, temp and moles
Exactly this! The volume is probably written on the container, so in terms of realism this is the more locigal calculation.

K-Dynamic commented:
I don't see why volumes need to be gatekept behind higher tech advanced gas analyzers or offscreen calculators

Sci has issues permeating tech across the station if they haven't blown up already

Offscreen calculators you either make yourself or ask someone via metacomms

Neither helps _new_ atmos techs who just joined the first time, get confused, then not play atmos again - which is why we keep seeing the same faces in atmos

> Exactly this! The volume is probably written on the container, so in terms of realism this is the more locigal calculation.

NT really needs to label nominal gas container properties (including volumes) on their tanks and canisters

Partmedia commented:
I'm still not convinced that gas analyzers need more information.

1. If the problem is really "The volume of tiles, pipes, tanks and canisters is not mentioned ingame" then the solution should be to add this to the examine text.

2. Because of the confusion between total volume for pipenets and cell volume for atmos tiles, this could add confusing and make it less accessible to new players.

3. This is more than "just QoL" because it gives more information than would have previously been available.

I would be happy to include this in an upgraded advanced gas analyzer, if you'd like to work on this. Otherwise I feel that the current gas analyzers are sufficient and the volume can be addressed with additional examine text, just like how real bottles are labeled with their capacity.

slarticodefast commented:
> If the problem is really "The volume of tiles, pipes, tanks and canisters is not mentioned ingame" then the solution should be to add this to the examine text.

I thought about that, too. But I think this might clutter the interface too much. Since the volume is constant, the information is not needed all the time like the tank pressure. You just have to look it up once for tanks.

> Because of the confusion between total volume for pipenets and cell volume for atmos tiles, this could add confusion and make it less accessible to new players.

I think this solves the unclarity that was there before. Before you had no way of knowing the if moles shown correspond to a cell volume or the total volume. Now you do.

> This is more than "just QoL" because it gives more information than would have previously been available.

It still is a one time calculation, that can be done on any volume that is not vacuum. But he ideal gas equation will still be used if you want to predict gas behaviour. For example, if I double the temperature, the pressure will get twice as high. Math will still be needed.

> I would be happy to include this in an upgraded advanced gas analyzer
I dont think that is a good solution for this. It would require a new research that would be barely used. Since my approach is to make it more accessible, the item should be available for new atmos players at round start and not locked behind research. And there would be two almost identical gas analyzers.

If you want another solution, I have the following suggestion:
- Show the volume of a single pipe element, tank, canister or map cell. This would be known in a real life setup as well since it should be written on the tanks or canisters or could be measured with a ruler.
- Show the temperature and pressure. This would be measured with a thermometer and a barometer.
- Only show the relative gas percentages, for example 3% plasma, 97% oxygen. This can be measured via mass spectrometry.
- Do not show the moles. There is no measurement device in real life which could count the total number of atoms in a pipe directly. If the player wants to know the number of moles, he has to use $pV=nRT$. This way the complexity stays the same. The moles are usually not needed, except for selling to cargo. For all gas reactions, only the relative percentages are important.

If only the relative percentages of gases are shown, the previous inconstistency with total and cell volume in the interface is fixed, while the player still has to apply math to get all information. The formula will basically just have to be solved for $n$ instead of $V$.

Partmedia commented:
> > If the problem is really "The volume of tiles, pipes, tanks and canisters is not mentioned ingame" then the solution should be to add this to the examine text.
> 
> I thought about that, too. But I think this might clutter the interface too much. Since the volume is constant, the information is not needed all the time like the tank pressure. You just have to look it up once for tanks.

I don't think it would unreasonably clutter the examine window. This would solve the most important issue, which is that people want a way to compare emergency tanks.

My preference for the solutions you proposed (1 being most preferable) are:

>     * Show the temperature and pressure. This would be measured with a thermometer and a barometer.
>     * Only show the relative gas percentages, for example 3% plasma, 97% oxygen. This can be measured via mass spectrometry.

This is what I personally find most preferable.

SlamBamActionman commented:
Just as a side observation; if the issue is the inability to properly convey the difference between a pipe system and a room tile, why not explicitly state it's a room tile in the gas analyzer UI? Could fit in the window title: 

"Gas Analyzer - Current Tile"
"Gas Analyzer - Pipe System"
"Gas Analyzer - Oxygen Tank"

Current Tile could maybe be made even more descriptive but oh well.

SlamBamActionman commented:
> > > If the problem is really "The volume of tiles, pipes, tanks and canisters is not mentioned ingame" then the solution should be to add this to the examine text.
> > 
> > 
> > I thought about that, too. But I think this might clutter the interface too much. Since the volume is constant, the information is not needed all the time like the tank pressure. You just have to look it up once for tanks.
> 
> I don't think it would unreasonably clutter the examine window. This would solve the most important issue, which is that people want a way to compare emergency tanks.
> 
> > ```
> > * Show the temperature and pressure. This would be measured with a thermometer and a barometer.
> > * Only show the relative gas percentages, for example 3% plasma, 97% oxygen. This can be measured via mass spectrometry.
> > ```
> 
> This is what I personally find most preferable. However, you might find a hard time selling this to the people who are used to the gas analyzer magically reporting the total number of moles :)

I do not see the point why any feature should be *locked* behind realism. While it is definitely a guiding principle in a simulation game, SS14 is clearly a game where the enjoyability and accessibility of a feature trumps realism. I see no point why there must be a real-life equivalent device to detect volume for this to be a thing, and it should primarily be argued for whether it is a relevant, useful and/or enjoyable addition to atmos gameplay.

We do not not have antag-detection glasses because it would be against realism, we do not have them because it'd be a shit game mechanic.

Partmedia commented:
> I do not see the point why any feature should be _locked_ behind realism. While it is definitely a guiding principle in a simulation game, SS14 is clearly a game where the enjoyability and accessibility of a feature trumps realism. I see no point why there must be a real-life equivalent device to detect volume for this to be a thing, and it should primarily be argued for whether it is a relevant, useful and/or enjoyable addition to atmos gameplay.
> 
> We do not not have antag-detection glasses because it would be against realism, we do not have them because it'd be a shit game mechanic.

I didn't suggest this for a reason. You suggested this.

Partmedia commented:
> I do not see the point why any feature should be _locked_ behind realism. While it is definitely a guiding principle in a simulation game, SS14 is clearly a game where the enjoyability and accessibility of a feature trumps realism. I see no point why there must be a real-life equivalent device to detect volume for this to be a thing, and it should primarily be argued for whether it is a relevant, useful and/or enjoyable addition to atmos gameplay.

Nobody is locking any feature behind realism. I think showing the total number of moles and total volume is:

- Bad for accessibility, because it confuses total moles when measuring a pipenet but cell moles in an environment
- Bad for gameplay, because disincentivizes independent exploration and discovery and turns the game into a spreadsheet numbers game
- Not enjoyable because number games encourage the kind of "hole up in atmos hyper-optimizing tritium" instead of the multiplayer dynamic gameplay we want
- Unnecessary because it could be done better by adding some examine text

deltanedas commented:
if gas analyzer can show moles it should be able to show volume too

> Bad for accessibility, because it confuses total moles when measuring a pipenet but cell moles in an environment

as said above this is already an issue and can be trivially fixed

> turns the game into a spreadsheet numbers game

without this thats literally what you have to do to figure out a room/net/tanks volume. pv=nrt or rtfsc and add up every things volume. sounds like spreadsheet game to me, even though needing volume is really niche. instead of using future spaceman analyzer that can do this automatically

hypothetically if moles were replaced with % you could use a known quantity of gas from rtfsc and pump it into a network then take away pv = nrt from the previous state and figure things out. it would only make atmos more spreadsheet gameplay

> Not enjoyable because number games encourage the kind of "hole up in atmos hyper-optimizing tritium" instead of the multiplayer dynamic gameplay we want

the people that use total volume are already in this crowd, and for everyone else atmos gameplay consists of going to the bar anyway

- Unnecessary because it could be done better by adding some examine text

examine text for pipes just removes a step of rtsfc you still have to go add everything up in your favourite calculator

for rooms either examining an empty tile just says its 2500L and nothing else since that doesnt do anything right now. nobody except atmos that are new will care about this so it would be bloat to *not* be in the analyzer

for entire pipenets why would some random tider be able to examine a pipe and just know its entire volume it makes even less sense than nagic spaceman tool

slarticodefast commented:
> "Gas Analyzer - Current Tile"
> "Gas Analyzer - Pipe System"
> "Gas Analyzer - Oxygen Tank"

This distinction is already made on the server side and would be easy to implement. I think this would be a good solution.
Another solution would be to just show the volume and moles of the selected pipe element, not the whole system. This would need some more calculation since internally the pipe network is stored as one big volume, not individual pipes.

> for rooms examining an empty tile just says its 2500L and nothing else since that doesnt do anything right now. nobody except atmos that are new will care about this so it would be bloat to not be in the analyzer

I agree. Empty tiles don't have an examination text and should not have one. So showing this on the gas analyzer would better.

> This is what I personally find most preferable. However, you might find a hard time selling this to the people who are used to the gas analyzer magically reporting the total number of moles :)

Is the exact number of moles actually needed somewhere, except maybe very precise maxcap crafting? I can making tritium and frezon just fine without ever looking at it. Only the pressure, temperature and gas percentages are needed. And if you want to know, you can still use the ideal gas equation to calculate it.
@Ilya246 maybe you can answer this.

> Bad for accessibility, because it confuses total moles when measuring a pipenet but cell moles in an environment

I still think showing the corresponding volume exactly fixes this problem.

Ilya246 commented:
> Is the exact number of moles actually needed somewhere, except maybe very precise maxcap crafting? I can make tritium and frezon just fine without ever looking at it. Only the pressure, temperature and gas percentages are needed. And if you want to know, you can still use the ideal gas equation to calculate it.
@Ilya246 maybe you can answer this.

not really, considering if you have read the source code you can simply just calculate how many moles of gas something contains, and maxcap recipes generally assume you're using a container of a specific volume which means the amount of moles of gas is known

slarticodefast commented:
@Partmedia I have now changed the gas analyzer so that it only shows the volume and corresponding moles of the selected pipe element. No more magic scanning of the whole station distro. Since this fixes the previous inconsistency I think this should be an acceptable solution. The video in the PR description has been updated and the changelog adjusted.
I will fix the merge conflicts as soon as possible.

Partmedia commented:
> @Partmedia I have now changed the gas analyzer so that it only shows the volume and corresponding moles of the selected pipe element. No more magic scanning of the whole station distro. Since this fixes the previous inconsistency I think this should be an acceptable solution. The video in the PR description has been updated and the changelog adjusted. I will fix the merge conflicts as soon as possible.

I think that's a fine middle-ground solution. Please go ahead and fix the merge conflicts and I'll double check with an in-game test before merging.

slarticodefast commented:
>It’s certainly sad to see the entire distro scan go. Maybe that behavior can be replicated in an advanced gas analyzer somewhere down the line

There is a atmosspheric monitoring console planned that gives you a overview over the entire pipe network, similar to the power monitoring console.

See #25938

> Not on this console; I plan to have the pipe net on its own console (the atmospheric monitoring console). This will also display atmos devices, such as gas pumps and thermo regulators, and the current statuses of these devices (e.g. if they are on or off, gas flow rates, set temperature)

slarticodefast commented:
Fixed the merge conflicts. Could someone double check this line change? #26493 changed it to use the ??= operator, which might have better performance. I undid this change because it was introducing an inconstency between the files GasFilterSystem.cs, GasMixerSystem.cs and GasVentPumpSystem.cs. All three have the OnAnalyzed function and it should be done the same way in each file.

![grafik](https://github.com/space-wizards/space-station-14/assets/161409025/4ee5af5e-f438-47b8-89da-a3dedfa79812)

I will also do some more in depth testing to make sure everything works as intended.

slarticodefast commented:
Added some optimizations suggested by @ElectroJr. These were partially added in #26493 and I made them consistent for all OnAnalyzed functions.
- Use ??= so that it actually supports scanning gases from an entity with multiple gas containing components.
- Replaced the dictionary in GasAnalyzerScanEvent with a list.
- Used the new NodeContainerSystem method that takes in a EntityUid instead of having to fetch the component separately.

Additional changes:
- The portable air scrubber no longer shows the port connector it is anchored to when scanned. This behaviour is now consistent with that of canisters.
- Canisters show inserted tanks when scanned.

On my side testing worked well and this PR is now ready for review.

Partmedia commented on Content.Server/Atmos/Piping/Unary/EntitySystems/GasVentPumpSystem.cs:
```diff
@@ -112,7 +112,7 @@ private void OnGasVentPumpUpdated(EntityUid uid, GasVentPumpComponent vent, ref
                     // Leak only a small amount of gas as a proportion of supply pipe pressure.
                     var pipeDelta = pipe.Air.Pressure - environment.Pressure;
-                    transferMoles = (float)timeDelta * pipeDelta * vent.UnderPressureLockoutLeaking;
+                    transferMoles = (float) timeDelta * pipeDelta * vent.UnderPressureLockoutLeaking;
```
Extraneous space

slarticodefast replied:
I only changed that because VSCodium was complaining. I saw the discussion on github, is the .editorconfig changed by now? I can remove the space, no problem.

Partmedia replied:
Looks like this is the only nitpick we're waiting on. Please also try to condense it down to two changelog entries if you can.

slarticodefast replied:
Both are done.

Partmedia commented on Content.Server/Atmos/Piping/Trinary/EntitySystems/GasMixerSystem.cs:
```diff
@@ -205,19 +205,31 @@ private void OnOutputPressureChangeMessage(EntityUid uid, GasMixerComponent mixe
-            if(_nodeContainer.TryGetNode(nodeContainer, component.OutletName, out PipeNode? outlet))
-                gasMixDict.Add(Loc.GetString("gas-analyzer-window-text-outlet"), outlet.Air);
+            // multiply by volume fraction to make sure to send only the gas inside the analyzed pipe element, not the whole pipe system
+            if (_nodeContainer.TryGetNode(uid, component.InletOneName, out PipeNode? inletOne) && inletOne.Air.Volume != 0f)
```
Is there a way to avoid copy/pasting the code from `[GasFilterSystem.cs](https://github.com/space-wizards/space-station-14/pull/25720/files#diff-7e8714bd40c3123783b5481b2a50536ca38c67ef7fedfe147efb3aa37e43ebb6)`?

slarticodefast replied:
GasFilterSystem.cs and GasMixerSystem.cs are similar in general, but I see no way to combine this without making them both inherit from a common trinary piping entity. I think changing that would be out of scope for this PR, other suggestions are welcome though.

Partmedia replied:
Understood, thanks.

Partmedia commented on Content.Server/Atmos/EntitySystems/GasTankSystem.cs:
```diff
@@ -359,7 +359,8 @@ public void CheckStatus(Entity<GasTankComponent> ent)
         private void OnAnalyzed(EntityUid uid, GasTankComponent component, GasAnalyzerScanEvent args)
         {
-            args.GasMixtures = new Dictionary<string, GasMixture?> { {Name(uid), component.Air} };
+            args.GasMixtures ??= new List<(string, GasMixture?)>();
```
Why are these assignments null-coalescing? Can `OnAnalyzed` now be called multiple times?

slarticodefast replied:
I asked @ElectroJr about this over Discord. This was changed by him in #26493, but only in GasFilterSystem.cs. He said the other files should be doing the same, so that the list does not get overwritten. His explanation was "The reason for changing it was just so that it actually supports scanning gases from an entity with multiple gas containing components". Im am not that familiar with how events work, but I think this is not really needed, but should it support this now in case more entities are added. At least this should be done in a consistent way in all OnAnalyzed functions, which it now does.

Partmedia replied:
Understood, thanks.

Partmedia reviewed (approved):
Thanks for your work, I've given this an in-game test and it looks good to go.

I left a few review comments if you're willing to address a few remaining nitpicks and questions, particularly about capturing the pattern of cloning the air mixture, multiplying to scale the volume, and then adding it to the output gas mixtures in a helper method.

slarticodefast commented on Content.Server/Medical/CryoPodSystem.cs:
```diff
@@ -276,10 +276,17 @@ private void OnGasAnalyzed(Entity<CryoPodComponent> entity, ref GasAnalyzerScanE
+        // multiply by volume fraction to make sure to send only the gas inside the analyzed pipe element, not the whole pipe system
+        if (_nodeContainer.TryGetNode(entity.Owner, entity.Comp.PortName, out PipeNode? port) && port.Air.Volume != 0f)
+        {
+            var portAirLocal = port.Air.Clone();
+            portAirLocal.Multiply(port.Volume / port.Air.Volume);
+            portAirLocal.Volume = port.Volume;
+            args.GasMixtures.Add((entity.Comp.PortName, portAirLocal));
```
I could add a function for this in [GasMixture.cs](https://github.com/space-wizards/space-station-14/blob/master/Content.Server/Atmos/GasMixture.cs). Although the existing functions in there are confusingly named. One could think Multiply(...), RemoveVolume(...), or RemoveRatio(...) would do what is needed here, but these do not adjust the volume. Also we need a clone since we do not want to modify the existing GasMixture in the pipe. Do you have a name suggestion for the new function which would make it clear that this one would work different from the others? MultiplyToVolume would be possible, but easy to confuse with Multiply(...) which modifies an existing GasMixture instead of returning a copy.

Partmedia replied:
Like you pointed out, we can get this in and look at this later. Thanks for looking at the options.

Partmedia commented:
Thanks for your contribution!
